### PR TITLE
UserSoty #18 Create business

### LINKED ISSUE #122: SP: 3 Implement view for Business creation
<b><i>User Story #18</b></i> 

<h2>Acceptance Criteria</h2>
Create View for Business where you can create its instance

### LINKED ISSUE #123: SP: 5 Create a serializer for Business POST method
<b><i>User Story #18</b></i> 

<h2>Acceptance Criteria</h2>
Create Serializer for CreateBusinessView where all required fields are listed, plus basic data validation, if needed

### LINKED ISSUE #124: - Add permissions (owner group) for CreateBusinessView
<b><i>User Story #18</b></i> 

<h2>Acceptance Criteria</h2>
Add permissions for users who can create business for himself

### LINKED ISSUE #125: SP: 3 Add tests for Business
<b><i>User Story #18</b></i> 

<h2>Acceptance Criteria</h2>
Add tests for CreateBussinessView, Serialization, and permissions.

### PR DESCRIPTION
**Pull request by morento101**
User Story #18 

Closes #123 
Closes #122 
Closes #124 
Closes #125

### CODE REVIEW

File: beauty/api/serializers/business_serializers.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ValidationError
+from django.utils.translation import gettext as _
+
+from api.models import Business
+
+User = get_user_model()
+
+
+class BusinessListCreateSerializer(serializers.ModelSerializer):
```
Comment (pazuzu-ua) [inline beauty/api/serializers/business_serializers.py:15]: please, add docstring

File: beauty/api/serializers/business_serializers.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    class Meta:
```
Comment (pazuzu-ua) [inline beauty/api/serializers/business_serializers.py:18]: please, add docstring

File: beauty/api/tests/test_business.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        response = self.client.post(
+            path=reverse("api:business-list-create"),
+            data=self.valid_create_data
+        )
+
+        self.assertEqual(response.status_code, 401)
+
+    def test_create_business_invalid_owner(self):
+        self.client = APIClient()
+        self.client.force_authenticate(user=self.test_client)
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_business.py:113]: No need to **force_authenticate** in every method. You can authentificate one time un setUp and then if you need to check an unauthentificated user, do **force_authenticate(user=None)**

File: beauty/api/tests/test_business.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        response = self.client.post(
+            path=reverse("api:business-list-create"),
+            data=self.invalid_create_data
+        )
+
+        self.assertEqual(response.status_code, 400)
+
+    def test_create_business_invalid_owner(self):
+        self.client = APIClient()
+        self.client.force_authenticate(user=self.test_client)
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_business.py:126]: No need to force_authenticate in every method. You can authentificate one time un setUp and then if you need to check an unauthentificated user, do force_authenticate(user=None)

File: beauty/api/tests/test_business.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        self.specialist2 = SpecialistFactory.create()
+
+        self.business = BusinessFactory.create()
+
+        self.position = PositionFactory.create()
+        self.position.specialist.add(self.specialist2)
+        self.business.position_set.add(self.position)
+
+    def test_create_position_method(self) -> None:
+        start_time, end_time = get_random_start_end_datetime()
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_business.py:41]: Should we check whether we have authority to create a position here?

File: beauty/api/tests/test_business.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    OwnerFactory, ClientFactory, SpecialistFactory, BusinessFactory, 
+    PositionFactory
+)
+from beauty.utils import get_random_start_end_datetime
+
+User = get_user_model()
+faker = Faker()
+
+
+class BusinessModelTest(TestCase):
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_business.py:19]: add docstring

File: beauty/api/tests/test_business.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    def setUp(self) -> None:
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_business.py:22]: add docstring

File: beauty/api/tests/test_business.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        self.specialist1 = SpecialistFactory.create()
+        self.specialist2 = SpecialistFactory.create()
+
+        self.business = BusinessFactory.create()
+
+        self.position = PositionFactory.create()
+        self.position.specialist.add(self.specialist2)
+        self.business.position_set.add(self.position)
+
+    def test_create_position_method(self) -> None:
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_business.py:37]: add docstring

File: beauty/api/tests/test_business.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        start_time, end_time = get_random_start_end_datetime()
+        position = self.business.create_position(
+            faker.word(), self.specialist1, start_time, end_time
+        )
+        all_positions = self.business.position_set.all()
+
+        self.assertIn(position, all_positions)
+        self.assertEqual(len(all_positions), 2)
+
+    def test_get_all_specialist_method(self) -> None:
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_business.py:50]: add docstring

File: beauty/api/tests/test_business.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        self.valid_create_data = {
+            "name": faker.word(), "type": faker.word(),
+            "description": faker.text(), "owner": self.owner.id
+        }
+        self.invalid_create_data = {
+            "name": faker.word(), "type": faker.word(),
+            "description": faker.text(), "owner": self.test_client.id
+        }
+
+    def test_list_of_businesses(self) -> None:
```
Comment (pazuzu-ua) [inline beauty/api/tests/test_business.py:89]: add docstring

File: beauty/api/permissions.py [hunk 11]
```diff
@@ -1,33 +1,30 @@
...
-    def has_object_permission(self, request, view, obj):
+    def has_permission(self, request, view):
         """Read permissions are allowed to any request,
         so we'll always allow GET, HEAD or OPTIONS requests.
         """
-        if request.method in permissions.SAFE_METHODS:
-            return True
-
-        return False
+        return bool(request.method in permissions.SAFE_METHODS)
```
Comment (Deer-WarLord) [inline beauty/api/permissions.py]: should be bool without explicit conversion

File: beauty/api/permissions.py [hunk 12]
```diff
@@ -1,33 +1,30 @@
...
     """
 
-    def has_object_permission(self, request, view, obj):
+    def has_permission(self, request, view):
         """Read permissions are allowed to any request,
         so we'll always allow GET, HEAD or OPTIONS requests.
         """
-        if request.method in permissions.SAFE_METHODS:
-            return True
+        return bool(request.method in permissions.SAFE_METHODS)
```
Comment (Deer-WarLord) [inline beauty/api/permissions.py]: should be bool without explicit conversion

File: beauty/api/permissions.py [hunk 13]
```diff
@@ -37,13 +34,14 @@ def has_object_permission(self, request, view, obj):
...
     """Object-level permission to only allow owners of an object to edit it."""
 
-    def has_object_permission(self, request, view, obj):
+    def has_permission(self, request, view):
         """Read permissions are allowed to any request,
         so we'll always allow GET, HEAD or OPTIONS requests.
         """
-        if request.method in permissions.SAFE_METHODS:
-            return True
+        return bool(
```
Comment (Deer-WarLord) [inline beauty/api/permissions.py]: should be bool without explicit conversion

File: beauty/api/permissions.py [hunk 14]
```diff
@@ -37,13 +34,14 @@ def has_object_permission(self, request, view, obj):
...
+            request.method in permissions.SAFE_METHODS or
+            request.user.is_authenticated
+        )
 
-        if request.user.is_authenticated and (obj.email == request.user.email):
-            return True
-        return False
+    def has_object_permission(self, request, view, obj):
+        return bool(obj.email == request.user.emai)
```
Comment (Deer-WarLord) [inline beauty/api/permissions.py]: should be bool without explicit conversion

**Review by Deer-WarLord — APPROVED** [review]

File: beauty/api/forms.py [hunk 15]
```diff
@@ -9,11 +9,13 @@ class CustomUserCreationForm(forms.ModelForm):
     """A form for creating new users. Includes all the required
     fields, plus a repeated password."""
     password1 = forms.CharField(label='Password', widget=forms.PasswordInput)
-    password2 = forms.CharField(label='Password confirmation', widget=forms.PasswordInput)
+    password2 = forms.CharField(label='Password confirmation',
```
Comment (therobotisnotatoy) [inline beauty/api/forms.py:12]: Priority is given to double quotes

File: beauty/api/forms.py [hunk 16]
```diff
@@ -9,11 +9,13 @@ class CustomUserCreationForm(forms.ModelForm):
...
+                                widget=forms.PasswordInput)
 
     class Meta:
         model = CustomUser
-        fields = ['email', 'first_name', 'last_name', 'patronymic', 'phone_number', 'bio', 'avatar']
+        fields = ['email', 'first_name', 'last_name', 'patronymic',
```
Comment (therobotisnotatoy) [inline beauty/api/forms.py:17]: Priority is given to double quotes

File: beauty/api/tests/test_business.py (hunk 5 above)
Comment (morento101) [reply beauty/api/tests/test_business.py:41]: This code tests the Business model method, not a view. Not sure that here we need to check any permissions or smth like this

Comment (pazuzu-ua) [reply beauty/api/tests/test_business.py:41]: You are right, sorry

**Pull request closed without merging**